Fortran-exponents

dhomeier opened the pull request:
My next highest priority after having a fast C reader was always being able to read in Fortran-written ascii files in its standard double-precision notation (`1.495978707D+13`…) without further processing,
which required setting up converters for each column with `np.loadtxt` or `np.genfromtxt`.
Taking a shot here at the fast_reader module to enable the c parser to directly recognise 'D' or other characters in the exponent.
This adds ‘fortran_exp’ as new option to the fast_reader arguments.

This patch automatically activates the fast_converter, as changing the call to xstrtod was
by far the simplest implementation. I guess it would be possible to add it to the more
precise parser as well, but don’t know how big the demand for this is. I have yet to see
a case where use_fast_converter gave a different result.
Asking for other exponent characters than 'E' therefore now turns use_fast_converter on by default; however setting it explicitly to False raises a FastOptionsError.

Notes following my initial post:
I have since noted that 'q' or 'Q' may be used in Fortran output as well, see post `#3` here
http://computer-programming-forum.com/17-c-language/9b704e6e49d734b8.htm

and therefore modified the option to `'fortran_exp': 'char'`

Implementing an ‘automatic’ option in `xtstrtod` I've experimented with a bit should be able to recognise any mix of `e`, `d`, `q` (or any other characters) as well as the triple-digit exponents mentioned in the link above, but with a probable performance penalty of 5-10 %. Also might increase the risk of converting strings to float64 one does not want to convert... Could probably try the auto-detection on the first lines and then stick with one format for each column, but that would require getting deeper into the entrails of `cparser.pyx`.

embray commented:
Something like this would be nice to have for #3430, since FITS ASCII tables use FORTRAN-inspired string formatting.

dhomeier commented:
If it's already using the c parser that should be fairly straightforward; if it isn't it would certainly benefit from the speedup as well. I'll have a look when I have time; I have some python-level code to check for the various exponent styles as well.

mdboom commented on astropy/io/ascii/cparser.pyx:
```diff
@@ -209,9 +210,17 @@ cdef class CParser:
             raise core.ParameterError("fast_reader cannot be False for fast readers")
-        # parallel and use_fast_reader are False by default
-        use_fast_converter = fast_reader.pop('use_fast_converter', False)
+        expchar = fast_reader.pop('fortran_exp', 'E').upper()
```
If I'm following along correctly, I think `fortran_exp` is used to specify the exponent character, but otherwise isn't really related to Fortran per se.  Maybe a better name would be `expchar`?

mdboom commented:
I think the automatic mode (where all of the reasonable options, `EDQedq`, are supported) may make some sense because it would lessen the need for the user option.  I think it's worth implementing and benchmarking to see how the performance compares.

Any sense of how common is the triple-digit exponent thing is?  I've never come across it (not that that means anything).

astrofrog commented:
The triple-digit exponent could happen if you computed for example blackbody fluxes in float64 as a function of frequency, since some values would be really small.

dhomeier commented:
Or chemical equilibrium fractions of compounds that don't really exist under the given conditions...
I actually have lots of data where those appear frequently, but they are small enough to be handled at the Python level.
The performance hit certainly would be tolerable, but kind of defeating the original purpose of the fast converter... Don't think it can be easily avoided even when not using the automatic option since the extra conditional(s) are in there in a rather time-critical loop.

mdboom commented:
I see -- so the triple-digit simplification happens automatically?  That's what I was getting at -- whether it was an automatic behavior or something a Fortran programmer would opt-in to.  Sounds like the former.

As for the potential performance hit: Let's do the most straightforward implementation, and benchmark it to get a sense of what the hit is.  If we can't deal with the hit, a solution would be to have multiple implementations of `xstrod`, one like it is now, and one with automatic behavior + triple-digit support.  When the user specifies "automatic mode" we move a function pointer to point to the latter.  Therefore the performance hit only goes into effect when automatic mode is on.  But we may not need to do all that -- it may be that other parts of the reader dominate over xstrtod anyway.

dhomeier commented:
Yes, the programmer can avoid it with an explicit format statement like `D20.12D3`, but by default the letter is just silently dropped.

OK, I've added the auto-detection with a bit of extra optimisation (thought the compiler would actually be able to abstract from the nested ifs...); for a real-world test case with some 2.3 Mio. lines, 26 columns 21 of which are read, maybe 20-40% of the original speed-up of use_fast_converter is lost (i.e. 7.02 s vs. 7.4 s have now increased to 7.12-7.2 s), but that's close to the timing uncertainty IMO. In fact the fastest runs have now occurred with the auto-detection turned on.

dhomeier commented:
I guess the way to add it to the standard parser would be to pre-process the entire string before passing it to `strtod`, similar to what's suggested [here](http://computer-programming-forum.com/47-c-language/021191896c8bfad9.htm):
```c
fgets(buf, sizeof buf, str);
for (p = buf; p = strchr(p, 'D'); *p = 'E') {} 
for (p = buf; p = strchr(p, 'Q'); *p = 'E') {}
```
(not tested nor seriously reviewed due to lack of C expertise!)

taldcroft commented:
@dhomeier - thanks, this is looking pretty good from a top-level look!  I'll try to go through in detail in the next couple days.  Perhaps @mdmueller can take a look as well.

I think just having this in the fast parser is reasonable.  Since the pure Python parser really just relies on numpy it would be a significant change to add this capability to the slow readers.

dhomeier commented:
@taldcroft - I rather had the not-so-fast C parser in mind anyway; for the Python parser it would mainly amount to providing a convenience function to specify converters like
`(n, lambda s: float(s.replace(b'D',b'e'))) for n in names`
but then, if users specified some converters of their own, the trouble might come in...

dhomeier commented:
Reviewing the build failures reported on AppVeyor and Travis, the former fails with the parallel version of `test_fortran_reader`, not surprising of course since the parallel reader is not supported under Windows. Just wondering if I had missed something in the setup of the `parametrize` decorator? The only test that specifically checks for the Windows case is `_read()`, yet none of the other tests seem to have reported failures either.

Travis build fails due to a Sphinx warning:
`/home/travis/build/astropy/astropy/build/lib.linux-x86_64-2.7/astropy/io/ascii/ui.py:docstring of astropy.io.ascii.read:98: WARNING: Block quote ends without a blank line; unexpected unindent.`
I did not remove any blank lines, don't know though if the 2nd-level indentation is illegal in a docstring.
It looks OK in `help(ascii.read)`, and seemed more structured to me this way...

taldcroft commented:
For the user API (the key in `fast_reader` dict), what about calling this `exponent_style` with following possible values:

- `E` (default), `D`, `Q` (case-insensitive): used as-is
- `fortran` (case-insensitive): same as the current `A` which allows any fortran-compatible style.

So not really any change to the C code, just a more obvious front-end interface.

taldcroft commented:
About the Sphinx warnings, you'll just need to play around until you get something that builds locally without warnings.  I'm often surprised by what Sphinx doesn't like.  I do like your idea of making it visually parseable.

taldcroft commented:
This change will also benefit from updating a couple of places in the narrative docs so people actually know about.  First is in the page about the fast reader:
  http://astropy.readthedocs.org/en/latest/io/ascii/fast_ascii_io.html

Next, I would suggest a new section called "Fortran-style exponents" just after the `Converters` section and at the same section header level as Converters:
  http://astropy.readthedocs.org/en/latest/io/ascii/read.html#converters

dhomeier commented:
Humm, running build_sphinx on my Mac returns a couple of warnings
__build succeeded, 2028 warnings.__
I'll try to watch out for those related to the new docstrings; I could see now that the html doc for `astropy.io.ascii.ui.read.html` did not render the 2nd indentation level right and apparently needed the blank line before that (seems to be the _Block quote ends without a blank line; unexpected unindent._ warning from Sphinx). Now the interactive help has an extra line, but probably can live with that.

I'll upload those doc fixes after changing the API to your suggestions. Any objections against silently accepting any substring of `Fortran` for the automatic mode?

taldcroft commented:
2028 warnings?  That's not normal...

dhomeier commented:
Don't know if the sphinx installation is broken - it's not terribly old, 1.2.3, but the python 3.4 variant does not work at all... Most of the warnings seem to be of this kind:
```
/Users/derek/git/astropy/docs/api/astropy.wcs.WCS.rst:40: WARNING: py:obj reference target not found: int
/Users/derek/git/astropy/docs/api/astropy.wcs.WCS.rst:76: WARNING: py:obj reference target not found: None
/Users/derek/git/astropy/docs/api/astropy.wcs.WCS.rst:82: WARNING: py:obj reference target not found: True
/Users/derek/git/astropy/docs/api/astropy.wcs.WCS.rst:89: WARNING: py:obj reference target not found: True
/Users/derek/git/astropy/docs/api/astropy.wcs.WCS.rst:115: WARNING: py:obj reference target not found: exceptions.ValueError
```

astrofrog commented:
That seems like errors that would happen if there was an issue with the internet connection. If you re-run the build, do you still get that many warnings?

dhomeier commented:
1633 - the connection should be fast enough, not sure if the firewall might cause any problems.

mdboom commented:
Near the beginning of the doc build it should say something like:

```
loading intersphinx inventory from http://docs.python.org/objects.inv...
loading intersphinx inventory from http://matplotlib.org/objects.inv...
loading intersphinx inventory from http://docs.scipy.org/doc/scipy/reference/objects.inv...
loading intersphinx inventory from http://docs.h5py.org/en/latest/objects.inv...
loading intersphinx inventory from http://docs.scipy.org/doc/numpy/objects.inv...
```

These are all http requests.  Are there any messages about them failing?  You may try removing the `docs/_build` directory before the rebuild in case any of those files are incomplete from an earlier failed request (that's just a shot in the dark, of course).

dhomeier commented:
@mdboom - thanks, you're right; they are failing:
```
loading intersphinx inventory from http://docs.python.org/objects.inv...
WARNING: intersphinx inventory 'http://docs.python.org/objects.inv' not fetchable due to <class 'urllib2.URLError'>: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:581)>
```
This is a "feature" of the latest openssl (at least the one in use on Mac OS since ~January); no idea if it is correct in not trusting certain certificate servers...
Not so easy to work around this, but I'll try to test the doc build on a Linux machine for the next commit.

dhomeier commented:
Just pinging - is there anything left to be done at this point, suggestions for labels (I missed to mark it as ENH, it's obviously `io.ascii`, no idea about how release-affecting this is)?
I guess some feedback/discussion on the documentation could be useful.

mdboom commented on astropy/io/ascii/src/tokenizer.c:
```diff
@@ -872,7 +877,45 @@ double xstrtod(const char *str, char **endptr, char decimal,
+            negative = 0;
+            switch (exp)
+            {
+            case '-': negative = 1;   // Fall through to increment pos
```
Can you put the statements on a new line, e.g:

```
case '-':
    negative = 1;
```

etc.

dhomeier replied:
I assume in the old code as well.

mdboom commented on astropy/io/ascii/src/tokenizer.c:
```diff
@@ -872,7 +877,45 @@ double xstrtod(const char *str, char **endptr, char decimal,
+                exponent += n;
+        }
+    }
+    else if (toupper(*p) == toupper(sci))
```
Maybe do `sci = toupper(sci)` outside of the inner loop?

dhomeier replied:
Before the `if (toupper(sci) == 'A')`?
Actually one option to pull the last bit of performance I had not played with! Since in the `! autoexp` case this is only needed this single time I thought it might be faster to call both `toupper()` right there in place, (I did experiment with `exp = toupper(*p);` outside of the loop), but the timing changes seem way below measurability. If there is really need for more speed-up, one should probably rely on the caller already passing the uppercase char (as it is already doing right now).
But probably not worth the loss in fault-tolerance.

mdboom commented on astropy/io/ascii/ui.py:
```diff
@@ -175,9 +175,18 @@ def read(table, guess=None, **kwargs):
+        use_fast_converter: bool
+            enable faster but slightly imprecise floating point conversion method
+        parallel: bool or int
+            multiprocessing conversion using ``cpu_count()`` or ``'number'`` processes
```
Maybe say:

```
Use multiprocessing to parse in parallel.  If `True`, use the number of cores on this machine.  If an `int`, indicates the number of cores to use.
```

mdboom commented on astropy/io/ascii/ui.py:
```diff
@@ -175,9 +175,18 @@ def read(table, guess=None, **kwargs):
-        (default= ``True``)
+        (default= ``True``); parameters for options dict:
+
+        use_fast_converter: bool
```
I don't think the nested argument formatting like this works.  I think you can just use a list of dictionary keys here, i.e.:

```
- ``use_fast_converter`` (bool): enable...
- ``parallel`` (bool or int): multiprocessing...
```

dhomeier replied:
This is processed to
<p><strong>fast_reader</strong> : bool or dict</p>
<blockquote>
<div><p>Whether to use the C engine, can also be a dict with options which default to <tt class="docutils literal"><span class="pre">False</span></tt>
(default= <tt class="docutils literal"><span class="pre">True</span></tt>); parameters for options dict:</p>
<dl class="docutils">
<dt>use_fast_converter: bool</dt>
<dd><p class="first last">enable faster but slightly imprecise floating point conversion method</p>
</dd>
<dt>parallel: bool or int</dt>
<dd><p class="first last">multiprocessing conversion using <tt class="docutils literal"><span class="pre">cpu_count()</span></tt> or <tt class="docutils literal"><span class="pre">'number'</span></tt> processes</p>
</dd>
<dt>exponent_style: str</dt>
<dd><p class="first last">One-character string defining the exponent or <tt class="docutils literal"><span class="pre">'Fortran'</span></tt> to auto-detect
Fortran-style scientific notation like <tt class="docutils literal"><span class="pre">'3.14159D+00'</span></tt> (<tt class="docutils literal"><span class="pre">'E'</span></tt>, <tt class="docutils literal"><span class="pre">'D'</span></tt>, <tt class="docutils literal"><span class="pre">'Q'</span></tt>),
all case-insensitive; default <tt class="docutils literal"><span class="pre">'E'</span></tt>, all other imply <tt class="docutils literal"><span class="pre">use_fast_converter</span></tt></p>
</dd>
</dl>
</div></blockquote>

which looks readable in Safari and Firefox (took the extra line at the top to make this work and stop Sphinx from complaining, which is a little bit annoying in `help()`, but I think one could live with it - see also taldcroft's comment above).
Would you suggest all keys in flowing text without line breaks, or did I misunderstand something?

taldcroft replied:
I looked at the built docs in chrome as well and this seems fine to me.

mdboom commented on astropy/io/ascii/ui.py:
```diff
@@ -175,9 +175,18 @@ def read(table, guess=None, **kwargs):
+        exponent_style: str
+            One-character string defining the exponent or ``'Fortran'`` to auto-detect
+            Fortran-style scientific notation like ``'3.14159D+00'`` (``'E'``, ``'D'``, ``'Q'``),
+            all case-insensitive; default ``'E'``, all other imply ``use_fast_converter``
```
```
all case-insensitive.  Default is ``'E'``, all others imply ``use_fast_converter=True``.
```

taldcroft commented on astropy/io/ascii/cparser.pyx:
```diff
@@ -209,9 +210,19 @@ cdef class CParser:
+            use_fast_converter = fast_reader.pop('use_fast_converter', True)
+            if not use_fast_converter:
+                raise core.FastOptionsError("fast_reader: exponent_style requires use_fast_converter")
+            if expchar.startswith('FORT'):
```
The documentation states that the parameter should be "Fortran", so check for equality to "FORTRAN" here.

dhomeier replied:
Bit of an Easter egg, yes - wasn't sure if we should encourage lazy typing here.

taldcroft commented on astropy/io/ascii/ui.py:
```diff
@@ -271,8 +280,18 @@ def _guess(table, read_kwargs, format, fast_reader):
+        parallel: bool or int
+            multiprocessing conversion using ``cpu_count()`` or ``'number'`` processes
+        exponent_style: str
+            Character to use for exponent or ``'Fortran'`` to auto-detect any
```
Since it is case-insensitive, might as well just say `'fortran'` here instead of `Fortran` just to make things simpler.

dhomeier replied:
Preferred the official spelling, I think it's clear enough that everything is case-insensitive, but I could keep that to the 'Fortran-style scientific notation' in the text of course.

taldcroft replied:
Well not a big deal if you like `Fortran`.  :smile:

pllim replied:
Shouldn't it be `FORTRAN`, since it is an acronym? :stuck_out_tongue_winking_eye:

dhomeier replied:
In "modern" style that would probably be `ForTran` then, for "Formula Translator", but the official name has gone to title-caps as of `Fortran 90` (roughly going along with not-ALL-UPPERCASE SOURCE CODE becoming generally tolerated :grimacing: ).

taldcroft commented on astropy/io/ascii/ui.py:
```diff
@@ -271,8 +280,18 @@ def _guess(table, read_kwargs, format, fast_reader):
+            multiprocessing conversion using ``cpu_count()`` or ``'number'`` processes
+        exponent_style: str
+            Character to use for exponent or ``'Fortran'`` to auto-detect any
+            Fortran-style scientific notation like ``'3.14159D+00'`` (``'E'``, ``'D'``, ``'Q'``),
```
This doesn't mention the 3-digit exponent style like `1.45e-123`.

taldcroft commented on docs/io/ascii/fast_ascii_io.rst:
```diff
@@ -84,6 +86,15 @@ IPython Notebook and so enabling multiprocessing in this context is discouraged.
+The special setting ``'Fortran'`` enables auto-detection of any valid
+Fortran-style exponent character like in ``'1.495978707D+13'``
+(``'E'``, ``'D'``, ``'Q'``), including triple-digit exponents prefixed
+with no character at all. 
```
Give an example of the triple-digit style.

taldcroft commented on docs/io/ascii/read.rst:
```diff
@@ -358,6 +358,20 @@ The default converters for each column can be overridden with the
   ...               'col2': [ascii.convert_numpy(np.float32)]}
   >>> ascii.read('file.dat', converters=converters)  # doctest: +SKIP
 
+
```
Include a cross reference link above the section header:
```
.. _fortran_style_exponents:
```

taldcroft commented on docs/io/ascii/fast_ascii_io.rst:
```diff
@@ -84,6 +86,15 @@ IPython Notebook and so enabling multiprocessing in this context is discouraged.
 slightly imprecise conversion method for floating-point values, as described below.
 
+The ``exponent_style`` parameter allows to define a different character
+from the default ``'e'`` for exponential formats in the input file.
```
I think you should cut the explanation below and just make a cross-reference here and say:
```
For details see the section on :ref:`fortran_style_exponents`.
```

taldcroft commented on docs/io/ascii/read.rst:
```diff
@@ -358,6 +358,20 @@ The default converters for each column can be overridden with the
+The :ref:`fast converter <fast_conversion_opts>` available with the C
+input parser provides an ``exponent_style`` option to define a custom
+character instead of the standard ``'e'`` for exponential formats in
+the input file, with the special setting ``'Fortran'`` for
```
Lower-case `'fortran'`.

taldcroft replied:
Give a few examples of valid exponential format styles.

taldcroft commented on astropy/io/ascii/tests/test_c_reader.py:
```diff
@@ -886,3 +886,41 @@ def test_fast_tab_with_names(parallel, read_tab):
+    assert_table_equal(table, expected)
+
+    # mixes and triple-exponents without any character using autodetect option
+    text = 'A B C\n1.0001+101 2.0E0 3\n.42d0 0.5 0.42-123'
```
Use `2.0-000` and `.42+000` for additional testing of the zero corner case.

taldcroft replied:
Test these values and make sure they end up as strings:
```
1.23-1
1.23+10
1.23+1002
1.23-100.2
```

dhomeier replied:
OK, that's probably something they won't (except for `1.23-100.2` I'd expect). More work on `tokenizer.c` to be done!

dhomeier replied:
Was not such a big deal after all...

taldcroft commented on astropy/io/ascii/src/tokenizer.c:
```diff
@@ -872,7 +877,45 @@ double xstrtod(const char *str, char **endptr, char decimal,
+        // check for possible Fortran exponential notations, including
+        // triple-digits with no character
+        exp = toupper(*p);
+        if (exp == 'E' || exp == 'D' || exp == 'Q' || *p == '+' || *p == '-')
```
Is there a reason to not use `exp` for the comparison with `'+'` and `'-'`?

dhomeier replied:
Merely that I did not know what `toupper('+')` would produce.

embray commented:
Likewise, how's this one looking?

taldcroft commented:
@embray - there are some open issues on this PR, awaiting action by @dhomeier.

dhomeier commented:
OK, I have no idea whatsoever is going on on Travis. Apparently the last (new) test in `test_fortran_reader()` has it hanging indefinitely when run with `parallel=True`, perhaps no so unexpected.

Marking it as xfail in the same manner as the other parallel reads now makes the test exit with error 245, whatever that is.

dhomeier commented:
Giving up on this for now, maybe someone with some insight into the Travis setup can provide a clue here. Apparently it's non-deterministic whether the parallel version of the last test in `test_fortran_reader()` is correctly raising an xfail, but evidently then the serial variant produces an error as well, whatever that error ("245") may be...

embray commented:
@dhomeier It looks like you did some kind of weird merge in this PR, and that's probably not helping.  You should go back and fix that like you did in #3667.

dhomeier commented:
I did what I did in #3667, i.e. (after an earlier frustrated rebase attempt)
```
git reset --hard 2db562d3be95538cd42bb6b74ed55f9576e09007
git rebase upstream/master fortranexp
```
which already produced dozens of merge conflicts for apparently trivial cases, apparently failed to update some files like `CHANGES.rst` entirely and left a whole subdirectory that had been removed from the repository behind.
Unless there is some compelling reason why it should work better on the third attempt, it seems to be less work now to pull a fresh branch from master and patch it myself rather then manually reviewing 50-100 merge conflicts another time. Besides I have no idea what another hard reset + forced push would do to this pull request...

dhomeier commented:
Oh well, going back to my last actual own commit (ba506f7ee818e1f01b9c948709d0a98f712ec640) made for a rather painless rebase; so if you think this is going to help with the state of this branch, I am willing to force-push it (looks like a large-scale deletion of history coming up, but I really don't care about that).

embray commented:
"looks like a large-scale deletion of history coming up, but I really don't care about that"  Not sure what you mean there.

But yeah if you've only made a few changes to a few files, if you do a rebase correctly you should not be getting hundreds of merge conflicts.  You should make sure that you're only rebasing your own changes--when you're trying to replay other commits from some other branch on top of the master branch (where they've probably already been merged) you'll end up with a mess.

dhomeier commented:
Well, there was a very different commit history in this PR until 30 minutes ago, but as I don't assume anyone would have gone through that anyway (or checked this branch out during that period), it should not matter if it's gone now.
Yes, it looks like the branch had already been halfway updated with base, whether automatically or accidentally by myself doing a push or pull too many I cannot tell anymore.
Anyway there don't seem to be any more problems with the merge as such, but the Travis test failures remain (and I did not really see why they would not, as the C reader sources and the test remained the same). But I've noticed there are a few more inconsistencies with very long exponents; going to look further into that first.

dhomeier commented:
In fact it might be worth discussing here what is the desired behaviour. I just added this test:
```
def test_data_out_of_range(parallel, read_basic):
    """
    Exponents beyond the float64 range (+308) should raise an error in
    the C parser and be returned as strings; below (-308) as float64 zeros.
    """
    text = 'a b c d\n100.01E+199 3.14e+313 2048e+306 0.6E-814'
    table = read_basic(text, parallel=parallel)
    expected = Table([['100.01E+199', '3.14e+313', '2048e+306', 0.0]],
                     names=('a', 'b', 'c', 'd'))
    assert_table_equal(table, expected)
```
which fails in the case where the standard reader is used (`t5` in the `_read` setup):
```
Reader = <class 'astropy.io.ascii.fastbasic.FastBasic'>, format = 'basic', parallel = False, check_meta = False, kwargs = {}
reader = <astropy.io.ascii.fastbasic.FastBasic object at 0x107bc6588>
t1 = <Table length=1>
     a          b         c        d    
  float64      str9      str9     str8  
----------- --------- --------- --------
1.0001e+201 3.14e+313 2048e+306 0.6E-814
t2 = <Table length=1>
     a          b         c        d    
  float64      str9      str9     str8  
----------- --------- --------- --------
1.0001e+201 3.14e+313 2048e+306 0.6E-814
t3 = <Table length=1>
     a          b         c        d    
  float64      str9      str9     str8  
----------- --------- --------- --------
1.0001e+201 3.14e+313 2048e+306 0.6E-814
t4 = <Table length=1>
     a          b         c        d    
  float64      str9      str9     str8  
----------- --------- --------- --------
1.0001e+201 3.14e+313 2048e+306 0.6E-814
t5 = <Table length=1>
     a         b       c       d   
  float64   float64 float64 float64
----------- ------- ------- -------
1.0001e+201     inf     inf     0.0
```
From @taldcroft's suggestions I take it returning strings is what we want to expect, so the non-fast reader would need to be fixed for that. The same is expected for very negative exponents, which the other exponent styles currently don't do (they return 0.0 for `"0.eD-814"`).

dhomeier commented:
Finally realised that the segfaults on Travis occur on the same cases as known from `test_data_out_of_range`, therefore marked it `xfail` know.

I note that there are still three different outcomes of too large exponents:
```
>>> text = 'a b c d\n10.1E+199 3.14e+313 2048e+306 0.6E-414'
>>> ascii.read(text, format='basic', guess=False, fast_reader=False)
<Table length=1>
    a        b       c       d   
 float64  float64 float64 float64
--------- ------- ------- -------
1.01e+200     inf     inf     0.0
>>> ascii.read(text, format='basic', guess=False, fast_reader={'exponent_style': 'e'})
<Table length=1>
    a         b         c        d    
 float64     str9      str9     str8  
--------- --------- --------- --------
1.01e+200 3.14e+313 2048e+306 0.6E-414
>>> ascii.read(text, format='basic', guess=False, fast_reader={'exponent_style': 'fortran'})
<Table length=1>
    a         b         c        d   
 float64     str9      str9   float64
--------- --------- --------- -------
1.01e+200 3.14e+313 2048e+306     0.0
```

Should one ultimately try to make the C reader return `inf` for too large exponents as well for consistency with the Python parser or even allow `float128` as return values (I imagine this should be feasible by modifying `xstrtod` to return exponent and mantissa as separate numbers)?

dhomeier commented:
One more thought, whatever the opinion on reading in extended-precision floats, the above listing shows one difference remaining between the existing `fast_reader` and the results using `exponent_style` extension: input with very small exponents is returned as `float64` rounded to zero rather than as string type.
Arguably the C parser's way of returning string arrays is, while not very pretty, for the time being still the most useful behaviour, since it at least allows one to repair the data pedestrian style after read-in:
```
newarr = np.empty(1, dtype=[('a', 'f8'), ('b', 'f16'), ('c', 'f16'), ('d', 'f16')])
newarr ['a'] = readarr['a']
newarr['b'][0] = np.float128(string(readarr['b'][0]))
...
```
So is there a general feeling it would be desirable to ensure consistency with the existing `fast_reader` functionality for the case of column 'd' above?

taldcroft commented on astropy/io/ascii/tests/test_c_reader.py:
```diff
@@ -888,3 +888,101 @@ def test_fast_tab_with_names(parallel, read_tab):
+    if TRAVIS:
+        pytest.xfail("Large exponents can sometimes fail on Travis CI")
+
+    text = 'A B C D E\n1.0001+1 2.3+10 3+1001 2 8.e3\n.42d0 0.5 3 4.56-123.4 0.42-123'
```
Testing with two rows is not so definitive because if either value parses as a string then the whole column is converted to string.  Better to write a clean, easy to read test like:
```
values = ['1.0001+1', '2.3+10', ..., '0.42-123']
expected = ['1.0001+1', '2.3+10', ..., 4.2e-122]
table = ascii.read(' '.join(values), format='no_header', ...)
out_values = [col[0] for col in table.itercols()]
assert out_values == expected
```

taldcroft commented on astropy/io/ascii/tests/test_c_reader.py:
```diff
@@ -888,3 +888,101 @@ def test_fast_tab_with_names(parallel, read_tab):
+    the C parser; such numbers are returned as strings instead, not so
+    with the Python parser!
+    """
+    if TRAVIS:
```
Can you make this `if TRAVIS and parallel`?  It seems like it should really work reliably if not doing parallel processing.

dhomeier replied:
I had only just changed it from `parallel and TRAVIS` in 4a44b6f2e752, pretty sure it did crash Travis in serial as well, but I could of course give it another try.

dhomeier replied:
I correct myself, seems that a couple of cases where `xstrtod` would return directly from the middle of the function without proper cleanup (likely returning a wild pointer in `endptr`...); looks like that's sorted out at last.

taldcroft commented on astropy/io/ascii/tests/test_c_reader.py:
```diff
@@ -888,3 +888,101 @@ def test_fast_tab_with_names(parallel, read_tab):
+    Make sure that ascii.read() can read Fortran-style exponential notation
+    using the fast_reader.
+    """
+    text = 'A B C\n100.01D+99 2.0 3\n4.2d-1 5.0D-1 0.6D4'
```
As above, better to make this one row.

taldcroft commented:
About the too-large exponents, for this PR I would suggest maintaining consistency with Python and returning `inf` or `0`.  Not all platforms support `float128` and in any case it would be better to have this discussion in a separate context.

taldcroft commented:
Overall it is difficult to look at the tests and get a good understanding of what is being tested, in particular with regard to what parses and what does not.  What would help is defining this in terms of a uniform test matrix like:
```
text_value expected exponent_style
...
```
You could actually do the same in code by just making a big list of tuples with those same three `value expected exponent_style`.  The point is that the coverage of various format possibilities (correct and incorrect) is crucial here.  It will be much more obvious when organized in a single table.

dhomeier commented:
OK, looks like the Travis errors are checked at last. The fast readers now return `inf` or `0` as `float64` for out_of_range values consistent with the Python parser, only the detailed results for some borderline cases may differ:
```
>>> asc.read('a\n 2.5e-324', format='basic', guess=False, fast_reader=False)['a']
<Column name='a' dtype='float64' length=1>
4.94065645841e-324
>>> asc.read('a\n 2.5e-324', format='basic', guess=False, fast_reader={'use_fast_converter': True})['a']
<Column name='a' dtype='float64' length=1>
4.94065645841e-324
>>> asc.read('a\n 2.5e-324', format='basic', guess=False, fast_reader={'use_fast_converter': False})['a']
<Column name='a' dtype='str8' length=1>
2.5e-324
>>> asc.read('a\n 2.45e-324', format='basic', guess=False, fast_reader={'use_fast_converter': False})['a']
<Column name='a' dtype='float64' length=1>
0.0
>>> asc.read('a\n 2.5e-308', format='basic', guess=False, fast_reader={'use_fast_converter': False})['a']
<Column name='a' dtype='float64' length=1>
2.5e-308
```
It seems for numbers larger than the smallest representable `float64`, `2.5e-324`, but smaller than `1/
np.finfo(np.float).max`, the builtin `strtod` (at least for the Darwin `gcc` here) raises `errno == ERANGE` without returning a value of `HUGE_VAL` or `0`. May need some more investigation, but I am beginning to wonder if we should trigger an `OVERFLOW_ERROR` in `cparser` at all, or just return `inf` or `0` in all those cases.
Will look into the tests structure later; the new `fast_reader` behaviour changed the outcome of `test_convert_overflow` as well, returning extremely long integer inputs as valid `float64` or `inf` now.

dhomeier commented:
Ah, I guess `strtod` raises the `ERANGE` flag because those numbers cannot be represented with the specified resolution of `1.e-15` (as is obvious above). But if the policy is not to fall back to `str` as long as possible, any inaccurate `float` would still be better than an even less accurate `0`, right?

taldcroft commented:
I obviously haven't dug into this at the level of detail you have been doing, but I think the guiding principle here should be predictability and consistency.  I.e. is there a simple way to predict what will happen for different cases, and is it consistent with behavior of Python and especially Numpy.  I'm not super worried about the extreme edge cases like 2.5e-324.

dhomeier commented:
I actually can't think of any case where the `ERANGE` flag should raise a `ValueError`, as all parsers now consistently return those numbers as `float64`. I think it would make sense to raise a `Warning` however, similar to the `OverflowError` warning issued by the `core.py` converter when the range of `IntType` is exceeded.

dhomeier commented:
BTW long integer conversion is still handled inconsistently between the Python parser, which switches right away to `str` when the range is exceeded, and the fast readers that try the `float` converter first. But in PR #2234 the agreement was eventually to fall back to `str` from `int`, so maybe that should be the common ground.
In any case, if the sequence of converters should remain `int` -> `float` -> `str`, it would be very desirable to have an option to override this to be able to read a column in string format at all. The `try_float` option in `cparser` unfortunately seems only settable from the file header.

dhomeier commented:
To summarise, I see 3 options for handling input values extending the respective numerical range:

1. `int` overflows are returned as `str` by the standard reader and promoted to `float` (which may include `inf`) by the fast readers; `float` inputs remain `float64`.

2.  `int` overflows are returned as `str` by the standard and fast readers; `float` inputs remain `float64`.

3.  `int` overflows are promoted to `float` (which may include `inf`) by all readers; `float` inputs remain `float64`.

*1 is the present situation, *2 would be the consistent solution keeping the spirit of #2234, so I'd tend to this option. In principle one could think of a 4th option always returning `float` overflows as `str` as well, but I don't see a straightforward way to identify overflows in the `convert_numpy` return, to begin with. I'd rather add some comments on explicitly specifying converters to return an `S` `dtype` to the documentation, if users want to preserve the full information.

taldcroft commented:
I would vote for option 2 as well.

dhomeier commented:
OK, it's been some time again before I got back to this; I have implemented option 2 from above
  `int` overflows are returned as `str` by the standard and fast readers; `float` inputs remain `float64`.
plus a `Warning` for the first overflow in either `int` or `float`, as well as possible. Some notes:
1. the actual fallback in mixed integer + float columns depends on order: if an `int` overflow shows up first, the column will be cast to `str`, if a float appears before that, it will be `float`. This same behaviour is already produced by the Python parser, and I don't see an easy way to fix it without reprocessing the entire column in question. I've put in an `xfail` test however to keep it on the radar.
2. The parser now ignores leading zeros in the input; working on that turned up that the integer parser was called with auto-detection flag set: `strtol(str, &tmp, 0)`, which caused it to interpret leading zeros as indicating an octal number. Depending on the digits this would either return the octal or raise a `CONVERSION_ERROR`. Given that it is quite unlikely anyone has used this feature, and it is also again inconsistent with the Python parser, I've changed the call to `strtol(str, &tmp, 10)`.

I've changed the invalid exponents tests according to taldcroft's suggestions, but left the multi-row input for some of the other tests, since columns with mixed types are part of the test case (see above).

However I feel rather lost with the new CI system - I cannot see any details at all on what's happening on AppVeyor, and CircleCI only shows this quite unhelpful message:

> ImportError: No module named jinja2

Also I don't find any way to investigate the merge conflicts reported for `astropy/io/ascii/tests/test_c_reader.py`...

dhomeier commented:
Could not find any way to work around the conflicts here, so I've merged the branch locally into the updated master and submitted this as a new PR #5552.
